Enable ETW/EventSource logging of task IDs for boxed state machines

delmyers opened the pull request:
Updates AsyncTaskMethodBuilderCore.TryGetTaskForContinuation so that it can return the task associated with the AsyncStateMachineBox.MoveNext function. This allows the TaskAwaiter to output the task id for the continuation when logging ETW.

Commit:
Wrap MoveNext Action for TPL event tracing

Wraps the MoveNext action of the AsyncStateMachineBox in a
continuation wrapper when async causality tracing is on so that the
TPL event source can find the task that is associated with a
continuation. Does not wrap otherwise.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/coreclr?pullRequest=27115) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/coreclr?pullRequest=27115) before we can accept your contribution.<br/><br/>:x: delmyers [sign now](https://cla.dotnetfoundation.org/dotnet/coreclr?pullRequest=27115)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/coreclr?pullRequest=27115) it.</sub>

delmyers commented:
This is a proposed fix for issue #27076

stephentoub commented on src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs:
```diff
@@ -284,13 +284,32 @@ private static void ExecutionContextCallback(object? s)
+                }
+
+                return moveNext;
+            }
```
@delmyers, can you please help me understand how this impacts the EventSource tracing?  I'm not seeing it.  CreateContinuationWrapper doesn't do any additional tracing.

delmyers replied:
Sure, it is because of this code in TaskAwaiter.cs:

```CSharp
            if (log.IsEnabled())
            {
                // ETW event for Task Wait Begin
                Task? currentTaskAtBegin = Task.InternalCurrent;

                // If this task's continuation is another task, get it.
                Task? continuationTask = AsyncMethodBuilderCore.TryGetContinuationTask(continuation);
                log.TaskWaitBegin(
                    currentTaskAtBegin != null ? currentTaskAtBegin.m_taskScheduler!.Id : TaskScheduler.Default.Id,
                    currentTaskAtBegin != null ? currentTaskAtBegin.Id : 0,
                    task.Id, TplEventSource.TaskWaitBehavior.Asynchronous,
                    continuationTask != null ? continuationTask.Id : 0);
            }
```

``AsyncMethodBuilderCore.TryGetContinuationTask(continuation)`` attempts to cast the continuation as a ``ContinuationWrapper`` to grab its associated task. So, when we create the continuation action from ``AsyncStateMachineBox<TStateMachine>.MoveNextAction, we have to make sure that it is wrapped in a ``ContinuationWrapper`` in order to allow ``TaskAwaiter.OutputWaitEtwEvents(…)`` to grab the associated task.

Note the comment that I added below, though. ``TaskAwaiter.OutputWaitEtwEvents(…)`` checks for ``TplEventSource.Log.IsEnabled()``, whereas ``AsyncStateMachineBox<T>.GetMoveNextAction()`` checks for ``AsyncCausalityTracer.LoggingOn``. These aren't exactly symmetrical. However, ``AsyncCausalityTracer.LoggingOn`` was what was used prior to Core 3 for wrapping the continuation, and the logging that it means to record is the ``TraceSynchronousWorkStart/Completion``, so that is what I used as well.

delmyers replied:
@stephentoub I've updated the comment to make it a little more clear.

stephentoub replied:
Thank you for the explanation.

I see now why your change addresses it.  However, it also adds non-trivial cost when logging is enabled, at least two allocations.  And while our logging-enabled path is more expensive than logging-disabled, we still want to try to minimize the expense, as we'd like developers to be able to enable logging with as little overhead as we can muster.

Given that, I'm not yet convinced it's the right fix.  For example, would it fix things to just change this:
https://github.com/dotnet/coreclr/blob/5b1c001bc0fb5bb1ce035c02ec275763f66defc8/src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilderCore.cs#L125-L126
to instead be this:
```C#
internal static Task? TryGetContinuationTask(Action continuation) =>
    continuation.Target is ContinuationWrapper wrapper ? wrapper._innerTask :
    continuation.Target as Task;
```
?

delmyers replied:
I could give that a try as well. My first instinct was to use the same scheme that was used previously, which was to wrap the continuation. I'll check to make sure that there aren't any other side-effects of doing it this way.

stephentoub replied:
Thank you.

delmyers replied:
@stephentoub, I've pushed the suggested changes to allow all of the automated tests to run. I'll let you know what my manual tests find as well.

Commit:
Clarifying comment

Commit:
removing trailing whitespace

delmyers commented:
@tarekgh @stephentoub It looks like a couple of the tests are failing, potentially due to a time-out. Can I get some help re-running the tests, or investigating the failures?

tarekgh commented:
I have restarted the failed tests.

Commit:
code review feedback.

Makes AsyncMethodBuilderCore.TryFindContinuationTask check to
whether the target of a continuation is itself a task, as a fall-back
to checking to see if the continuation is a ContinuationWrapper

stephentoub commented on src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilderCore.cs:
```diff
@@ -123,7 +123,9 @@ internal static string GetAsyncStateMachineDescription(IAsyncStateMachine stateM
-            (continuation?.Target as ContinuationWrapper)?._innerTask;
+            (continuation?.Target is ContinuationWrapper wrapper) ?
+                wrapper._innerTask :            // A wrapped continuation, created by an awaiter
+                continuation?.Target as Task;   // The continuation targets a task directly, such as with AsyncStateMachineBox
```
Looking at all the call sites, I don't believe it's possible for `continuation` to ever be null here, in which case the `?`s on `continuation?.` can be removed in both places.  But it's fine to not do so as part of this change.

delmyers replied:
This is strange... by a applying this change, I'm now missing some operation started events. It doesn't make any sense, unless there are more issues with dotnet 5.

By the way, I have a suite of unit tests that I would like to execute these changes against to see what the impact is. vstest uses the installed version of core 3 to run the tests. Do you know of a way that I can patch my machine to make vstest run dotnet 5 with my changes instead?

delmyers replied:
Never mind, I got it working. My local tests so far indicate that this will work as well.

Commit:
Got rid of unnecessary null checks

delmyers commented:
@stephentoub @tarekgh, so all of the clr tests pass. I've also checked this against a bunch of local unit tests, and my unit tests all pass. Do you think that I could get a sign-off on this change?

tarekgh commented:
@stephentoub if you don't have any further comments, can we merge this?

stephentoub reviewed (approved):
Thanks.
